pymc_experimental/model_transform.py

```diff
@@ -0,0 +1,175 @@
+        import pymc as pm
+        from pymc_experimental.model_transform import uncensor
+
+        with pm.Model() as model:
+            x = pm.Normal("x")
+            dist_raw = pm.Normal.dist(x)
+            y = pm.Censored("y", dist=dist_raw, lower=-1, upper=1, observed=[-1, 0.5, 1, 1, 1])
+            idata = pm.sample()
+
+        with uncensor(model):
+            idata_pp = pm.sample_posterior_predictive(idata, var_names=["y_uncensored"])
+
+        az.summary(idata_pp)
+    """
+    fg = fgraph_from_model(model)
+
+    (_, nodes_changed, *_) = uncensor_rewrite.apply(fg)
+    if not nodes_changed:
+        raise RuntimeError("No censored variables were replaced by uncensored counterparts")
+
+    return model_from_fgraph(fg)
+
+
+@node_rewriter(tracks=[ModelValuedVar])
+def forecast_timeseries_node_rewrite(fgraph: FunctionGraph, node):
+    """Rewrite that replaces censored variables by uncensored ones"""
+
+    (
+        timeseries_rv,
+        value,
+        *dims,
+    ) = node.inputs
+    if not isinstance(timeseries_rv.owner.op, AutoRegressiveRV):
+        return
+
+    forecast_steps = get_var_by_name(fgraph.inputs, "forecast_steps_")
+    if len(forecast_steps) != 1:
+        return False
+
+    forecast_steps = forecast_steps[0]
+
+    op = timeseries_rv.owner.op
+    model_rv = node.outputs[0]
+
+    # We cannot reference the variable we are planning to replace
+    # Or it will introduce circularities in the graph
+    # FIXME: This special logic shouldn't be needed for ObservedRVs
+    # but PyMC does not allow one to not resample observed.
+    # We hack around by conditioning on the value variable directly,
+    # even though that should not be part of the generative graph...
+    if isinstance(node.op, ModelObservedRV):
+        init_dist = DiracDelta.dist(value[..., -op.ar_order :])
+    else:
+        cloned_model_rv = model_rv.owner.clone().default_output()
+        fgraph.add_output(cloned_model_rv, import_missing=True)
+        init_dist = DiracDelta.dist(cloned_model_rv[..., -op.ar_order :])
+
+    if isinstance(timeseries_rv.owner.op, AutoRegressiveRV):
+        rhos, sigma, *_ = timeseries_rv.owner.inputs
+        new_timeseries_rv = AR.rv_op(
+            rhos=rhos,
+            sigma=sigma,
+            init_dist=init_dist,
+            steps=forecast_steps,
+            ar_order=op.ar_order,
+            constant_term=op.constant_term,
+        )
+
+    new_name = f"{model_rv.name}_forecast"
+    new_value = new_timeseries_rv.type(name=new_name)
+    new_timeseries_rv = model_free_rv(new_timeseries_rv, new_value, transform=None)
+    new_timeseries_rv.name = new_name
+
+    # Import new variables into fgraph (value and RNG)
+    fgraph.import_var(new_timeseries_rv, import_missing=True)
+
+    return [new_timeseries_rv]
+
+
+forecast_timeseries_rewrite = in2out(forecast_timeseries_node_rewrite, ignore_newtrees=True)
+
+
+def forecast_timeseries(model: Model, forecast_steps: int) -> Model:
```

Comment:
We could maybe request a `replace_mutable_dims={"old_dim": "new_dim"}`. Right now if one has dimmed deterministics or variables below the timeseries they will still correspond to something potentially no longer valid.

Reply:
I had the GP predictions in mind with this example, but this was a bit simpler as all the graph information is available in the model. For more context see https://gist.github.com/ricardoV94/c16693680d55294eb73feed7c9997421

Reply:
Gotcha, still have your gist up in a tab..  I'm thinking that a really nice way to conceptualize GPs, or AR models like this, would just be as functions.  

```
with pm.Model() as model:
    X = pm.MutableData("X", X_)
    gp = pm.gp.Latent(cov_func=cov_func)
    f = gp(X)  # formerly gp.prior(...)
    ...
    pm.sample()


with model:
    pm.set_data(X_new)
    # or maybe pm.set_data({"X": X_new}, replace_dims={"X": new_coords})

    
```
Is all that's needed is to tuck your function [here](https://gist.github.com/ricardoV94/c16693680d55294eb73feed7c9997421) into `pm.set_data`?

Reply:
I think we want to create a new model and not mess with the original one. In these cases the predictive model is different than the prior model, which is not such an unreasonable thing.

Reply:
Ah right, it'd be nice from a user perspective though if there was one way to do this regardless of whether its a linear model, AR, or GP.  

My bad, obv you cant return a new model somehow within `pm.set_data`.  What if the syntax switched a bit, `new_model = pm.set_data(X_new)` (or a new function, `pm.update_data`) and could handle the diff cases?

Reply:
What if you have a Censored Timeseries model with a GP? Are we assuming the user wants to uncensor, forecast and condition on GP predictions all at the same time? What if they only want one of these?

With the approach here it's easy, just chain function calls:
```python
with uncensor(forecast_timeseries(model)) as pred_model:
  pm.set_data({"x": new_data})  # If you need to
  pm.sample_posterior_predictive(...)
```

For creating a predictive model we might also need some arguments, like `forecast_steps` for the timeseries case here
or `new X` for GP. A generic function would have to receive all the arguments, which would become difficult for users to figure out. With specific functions you can just read their docstrings.

Also `set_data` can be used completely outside of the realm of posterior predictive. To update your model observations, or perform prior_predictive studies.

Some users equate `set_data` with getting predictions, but that's just one specific strategy that works for a narrow range of models. I don't think it will take much training to break that association. `set_data` != `make_predictions`. WDYT?

Reply:
```python
with uncensor(forecast_timeseries(model, forecast_steps=10)) as pred_model:
    pm.set_data({"x": new_data})  # If you need to
    pm.sample_posterior_predictive(...)
```
^ This is super slick.  But I think part of what makes PyMC nice is you can do both right?  A function like `pm.sample` has tons of arguments too, but the defaults work great for 80% of use cases. And if you wan't you can call the functions `pm.sample` calls, so you can have it both ways.  I admit I think of `set_data` as for predictons in my head, it's def more broadly useful.  

Def take your point about a censored timeseries with a GP.  It's great how you did it.  But, it's pretty common to have a linear model `X @ beta` and one or many GPs together in the same mode.  It'd be nice if there's a way to update `X` and have that propagate without updating each individual component, potentially different ways.  Maybe it just boils down to whether `set_data` interacts with GPs?

Reply:
The API I have in mind for GPs would be:

```python
with predict_gp(gp1=x1, gp2=x2, ...) as pred_m:
  pm.sample_posterior_predictive()
```

That feels much more intuitive. 

Procedurally `set_data` doesn't seem like good analogy for GPs. We don't want to override the initial x, we want to expand them (and apply a specific multivariate normal to that expansion). If it was called `pm.expand_data` it would be closer, but again I don't think it's worth putting a square through a circle.

I think sample is actually a good counter example, that code is really hard to understand and maintain. Users have no idea how to do something less standard like resuming sampling which libraries like blackjax and numpyro do much better.

Also a more practical if silly reason is that we can't create a distinct model inside the context of another. The variables would be added to the parent. And you can't call `set_data` outside of a model.

But instead of me arguing against it, do you see a problem with this approach? Do you think it would be hard to use, to find or both?

Reply:
How does the `gp1=x, gp2=x2, ...` part work? is that setting X for each gp object in the model?

The details of predicting with a GP could be abstracted away though.  It'd be nice if a user could just think of GPs as random functions `f(x)` that take some `x` and return a distribution.  So you sample initially with some `x`, then you can plug in new `x` and get the new GP at x.  Which is true, the stuff about conditionals and non-parametric and all that is kind of an implementation detail, and there to make working with GPs more efficient.  It's just as correct to fit the whole thing with NUTS, a GP over x (where data is observed) and x_new (where it isnt) up front.  It's too slow though.

One potential problem (don't know if it necessarily is), is how this would interact with [additive GPs](https://www.pymc.io/projects/docs/en/latest/guides/Gaussian_Processes.html#additive-gps), example [here](https://www.pymc.io/projects/docs/en/v3/pymc-examples/examples/gaussian_processes/GP-MaunaLoa.html#Examining-the-fit-of-each-of-the-additive-GP-components).  Could you apply this to predict a component of an additive GP?  (I think its def possible, `predict_gp` might need some args for this, just wanted to keep it in mind as a potential complication)

It also kind of looks like you're building higher level model components, so censoring, gps, timeseries, which all use this function chaining for uncensoring/prediction/forecasting whatever is natural.  Even though you could do it with `pm.set_data`, you could additional build a `predict_lm` function that works the same more function chaining way too.   

If you had a linear model as the mean of a GP, or equiv., $y \sim N(f + X\beta, \sigma^2)$, would you do 
```python
with predict_gp(...) as pred_m:
  pm.set_data({"X": x_new})
  pm.sample_posterior_predictive()
```
(or maybe, this one?)
```python
with predict_gp(predict_lm(...)) as pred_m:
  pm.sample_posterior_predictive()
```

Anyway, please don't let this discussion be a drag!  I feel there's a few threads here and theres a lot to think about.

Reply:
> One potential problem (don't know if it necessarily is), is how this would interact with [additive GPs](https://www.pymc.io/projects/docs/en/latest/guides/Gaussian_Processes.html#additive-gps), example [here](https://www.pymc.io/projects/docs/en/v3/pymc-examples/examples/gaussian_processes/GP-MaunaLoa.html#Examining-the-fit-of-each-of-the-additive-GP-components). Could you apply this to predict a component of an additive GP? (I think its def possible, predict_gp might need some args for this, just wanted to keep it in mind as a potential complication)

I have to read a bit about that to understand it.

> If you had a linear model as the mean of a GP, or equiv., , would you do

Either would be possible! The `GLM` part usually doesn't need any graph transformation (that's why it already works). Although you could use it replace ConstantData predictors by a new one if you wanted to. This way a PyMC model can sample with constant predictors, which may allow more efficient graphs, than when we use shared variables where it has to assume those could change values/shape any time.

But I would probably go for the first (not implement a `predict_lm` transform). The `predict_gp`, creates a new model with the prediction component inserted right after where the original `gp` was (conditioned on the new_X). After that you can still set any old `MutableData` variables you had in the original model as you want.

Reply:
> Anyway, please don't let this discussion be a drag! I feel there's a few threads here and theres a lot to think about.

This was the point of me opening the PR, so not a drag at all. The idea is to experiment and see if something works well enough. Maybe I should add a `gp_predict` transform in this PR to see how it looks like?

Reply:
okay definitely, I'm curious how it looks.  

> The predict_gp, creates a new model with the prediction component inserted right after where the original gp was (conditioned on the new_X). After that you can still set any old MutableData variables you had in the original model as you want.

Would mutating `X` affect the original GP though?  Like, if you have 
```python
f = gp.prior("f", X)
...
pm.set_data({"X": x_new})
```
you might expect something to happen to `f`?